File: 66-plus-one/66-plus-one.java
class Solution {
    public int[] plusOne(int[] d) {
        int n = d.length;
        if (d[n - 1] != 9) { d[n - 1] += 1; return d; }
        for (int i = n - 1; i >= 0; --i) {
            if (d[i] == 9) {
                d[i] = 0;
            } else {
                d[i] += 1;
                return d;
            }
        }
        int[] extra = new int[n + 1];
        extra[0] = 1;
        for (int i = 0; i < d.length; ++i) extra[i + 1] = d[i];
        return extra;
    }
}

File: 66-plus-one/66-plus-one.cpp
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int n = digits.size();
        if (digits[n - 1] != 9) { digits[n - 1]++; return digits; }
        reverse(digits.begin(), digits.end());
        digits[0] = 0;
        for (int i = 1; i < n; ++i) {
            if (digits[i] == 9) digits[i] = 0;
            else {
                digits[i]++;
                break;
            }
        }
        if (digits[n - 1] == 0) digits.push_back(1);
        reverse(digits.begin(), digits.end());
        return digits;
    }
};